File: src/main/java/jaba/lists/linkedlists/singlelinkedlists/JosephusProblem.java
package jaba.lists.linkedlists.singlelinkedlists;

import jaba.lists.linkedlists.singlelinkedlists.recursive.Node;

/**
 * Created by Administrador on 22/05/2017.
 */
public class JosephusProblem {

    final int romanSoldiersNumber;
    final int iteration;

    @java.beans.ConstructorProperties({"romanSoldiersNumber", "iteration"})
    public JosephusProblem(int romanSoldiersNumber, int iteration) {
        this.romanSoldiersNumber = romanSoldiersNumber;
        this.iteration = iteration;
    }

    Node<Integer> getSurvivor() {
        Node<Integer> tail = new Node<>(0);
        Node<Integer> survivor = tail;
        for (int i = 1; i <= romanSoldiersNumber; i++) {
            survivor.setNext(new Node<>(i));
            survivor = survivor.getNext();
        }
        survivor.setNext(tail);
//        while (!survivor.equals(survivor.getNext())) {
        while (survivor != survivor.getNext()) {
            for (int i = 0; i < iteration; i++) {
                survivor = survivor.getNext();
                survivor.setNext(survivor.getNext().getNext());
            }
        }
        return survivor;
    }

    public static void main(final String[] args) {
        final JosephusProblem josephusProblem = new JosephusProblem(10, 10);
        final Node<Integer> survivor = josephusProblem.getSurvivor();
        //log.info(survivor.toString());
    }
}

File: src/main/java/katlin/linkedlists/singlelinkedlists/JosephusProblem.kt
package katlin.linkedlists.singlelinkedlists

import katlin.linkedlists.singlelinkedlists.recursive.Node

/**
 * Created by Administrador on 24/05/2017.
 */
class JosephusProblem(internal val romanSoldiersNumber: Int, internal val iteration: Int) {

    internal val survivor: Node
        get() {
            val tail = Node(1)
            var survivor = tail
            for (i in 0..romanSoldiersNumber - 1) {
                survivor.next = Node(i)
                survivor = survivor.next!!
            }
            survivor.next = tail
            while (survivor != survivor.next) {
                for (i in 0..iteration - 1) {
                    survivor = survivor.next!!
                    survivor.next = survivor.next!!.next
                }
            }
            return survivor
        }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
                val josephusProblem = JosephusProblem(10, 10)
                System.out.println(josephusProblem.survivor)
        }
    }
}